### PR TITLE
openshift: allow custom payload images for development

### PR DESCRIPTION
**Pull request by jensfr**
Sometimes we need to test new builds of RPMs, for example to
verify a bug in QEMU, kata-runtime or other packages.

For this purpose the user can now add a configmap with a single data
entry that specifies a URL to a container image, e.g.
quay.io/someuser/repository:tag.

The ConfigMap will be added to the Pod spec of the installer daemon pods as an
environment variable. If the variable is set the daemon will use it. If
not the default mechanism for choosing the payload image is used.

The purpose of this feature is for development purposes only. When the
ConfigMap is used the operator will print a warning that using
self-built RPMs taints the kataconfig installation.

Also add a description in docs/HACKING.md on how to use the ConfigMap
and how to build a custom payload image.

Signed-off-by: Jens Freimann <jfreimann@redhat.com>

### CODE REVIEW

**Comment by jensfr** [comment]
@zanetworker

**Comment by harche** [comment]
@jensfr nice :)

File: doc/HACKING.md [hunk 1]
```diff
@@ -0,0 +1,27 @@
+# Hacking on the kata-operator
```
Comment (zanetworker) [inline doc/HACKING.md:1]: I thinking Hacking is cool in the content of the doc, but for the title (also the first doc we add :D) maybe `local-development` / `development` or something along would fit more? WDYT?

File: doc/HACKING.md [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+## Using a custom kata-operator-payload image
+
```
Comment (zanetworker) [inline doc/HACKING.md:4]: Do you think it makes sense to add the "why" (probably copied from the issue description) we would want to set a custom image?

**Comment by jensfr** [comment]
Just realized the daemon is also part of this repo now and I can include the change in this PR. Will update the branch soon.

File: pkg/controller/kataconfig/openshift_reconciler.go [hunk 3]
```diff
@@ -162,6 +162,7 @@ func (r *ReconcileKataConfigOpenShift) isOldestCR() (bool, error) {
 
 func (r *ReconcileKataConfigOpenShift) processDaemonsetForCR(operation DaemonOperation) *appsv1.DaemonSet {
 	runPrivileged := true
+	configmapOptional := true
```
Comment (zanetworker) [inline pkg/controller/kataconfig/openshift_reconciler.go:165]: maybe group all the vars under a var block: 

example: 
```go
var (
	runPrivileged = true
	configmapOptional = true
	var runAsUser int64 = 0
)
```

File: deploy/configmap_payload.yaml [hunk 4]
```diff
@@ -0,0 +1,8 @@
+kind: ConfigMap
+apiVersion: v1
+metadata:
+  creationTimestamp: 2016-02-18T19:14:38Z
+  name: payload-config
+  namespace: kata-operator
+data:
+  daemon.payload: quay.io/jensfr/kata-operator-payload:test
```
Comment (zanetworker) [inline deploy/configmap_payload.yaml:8]: maybe add an indication that the value is an image tag?

File: doc/HACKING.md [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+Note: it is not possible to add additonal RPMs this way
+
+An example:
+
+1. skopeo copy docker://quay.io/jensfr/kata-operator-payload:4.6.0 oci:/tmp/kata-operator-payload:4.6.0
+2. oci-image-tool unpack --ref name=4.6.0  /tmp/kata-operator-payload kata-operator-payload-unpacked-4.6.0
+3. cp -r /tmp/kata-operator-payload-unpacked-4.6.0/packages $KATA_OPERATOR_REPO/images/payload
+4. cd $KATA_OPERATOR_REPO/images/payload
+5. replace RPMs in packages/ with custom RPMs
+6. podman build --no-cache -f Dockerfile.custom quay.io/<username>/mykatapayload:mytag
```
Comment (zanetworker) [inline doc/HACKING.md:24]: can it also be built with docker?

**Comment by zanetworker** [comment]
One other thing, are you also planning to cherry pick this master?

File: images/daemon/pkg/daemon/kata_openshift.go [hunk 6]
```diff
@@ -406,9 +406,18 @@ func installRPMs(k *KataOpenShift) (err error) {
 		fmt.Println(err)
 	}
 
-	srcRef, err := alltransports.ParseImageName("docker://quay.io/isolatedcontainers/kata-operator-payload:" + k.PayloadTag)
+	payloadImage := os.Getenv("KATA_PAYLOAD_IMAGE")
+	if payloadImage == "" {
+		payloadImage = "docker://quay.io/isolatedcontainers/kata-operator-payload:" + k.PayloadTag
+	} else {
+		log.Println("WARNING: kataconfig installation is tainted")
```
Comment (zanetworker) [inline images/daemon/pkg/daemon/kata_openshift.go:413]: What is meant by tainted here?

File: images/daemon/pkg/daemon/kata_openshift.go [hunk 7]
```diff
@@ -406,9 +406,18 @@ func installRPMs(k *KataOpenShift) (err error) {
...
+		log.Println("Using env variable KATA_PAYLOAD_IMAGE " + payloadImage)
+		payloadImage = "docker://" + payloadImage
+	}
+
+	srcRef, err := alltransports.ParseImageName(payloadImage)
 	if err != nil {
-		fmt.Println("Invalid source name")
+		fmt.Println("Invalid source name of payload container image: " + payloadImage)
```
Comment (zanetworker) [inline images/daemon/pkg/daemon/kata_openshift.go:420]:
```suggestion
		fmt.Printf("Invalid source name of payload container image: %q", payloadImage)
```

File: doc/HACKING.md (hunk 1 above)
Comment (jensfr) [reply doc/HACKING.md:1]: I thought the name was common in related projects. machine-config-operator has the same file, opencontaimer/image-spec, cluster-etc-operator all have a HACKING.md. I wanted not to deviate from that. But that's the only reason. I have no strong opinion either way, so I'll just name it DEVELOPMENT. Ok?

File: doc/HACKING.md (hunk 2 above)
Comment (jensfr) [reply doc/HACKING.md:4]: It sure doesn't hurt :-) :+1:

File: doc/HACKING.md (hunk 5 above)
Comment (jensfr) [reply doc/HACKING.md:24]: what's docker? ;-) Yes, I  see no reason why not.

File: images/daemon/pkg/daemon/kata_openshift.go (hunk 6 above)
Comment (jensfr) [reply images/daemon/pkg/daemon/kata_openshift.go:413]: Same meaning it has when you load a proprietary kernel module. You're on your own when problems occur. 
https://www.kernel.org/doc/html/latest//admin-guide/tainted-kernels.html

File: pkg/controller/kataconfig/openshift_reconciler.go (hunk 3 above)
Comment (jensfr) [reply pkg/controller/kataconfig/openshift_reconciler.go:165]: :+1:

File: doc/HACKING.md (hunk 1 above)
Comment (zanetworker) [reply doc/HACKING.md]: Sure, the reason I commented is that other OS projects have a dedicated folder usually called `hack` that contains scripts for many things. But it is up to you, no strong opinions on my side as well :)

Comment (jensfr) [reply doc/HACKING.md]: ah, I did not get that from your comment. Sure, I think we add a helper script in hack/ to build a custom payload but I'd prefer to do that in another PR.

**Comment by jensfr** [comment]
> One other thing, are you also planning to cherry-pick this to master?

Yes, I'll send a PR for master as well but it requires some small changes.

File: doc/HACKING.md (hunk 1 above)
Comment (zanetworker) [reply doc/HACKING.md]: It looks good as it is now :)

**Pull request merged**